lib/completion.ts

```diff
@@ -34,6 +34,78 @@ export function completion(
     const current = args.length ? args[args.length - 1] : '';
     const argv = yargs.parse(args, true);
     const parentCommands = yargs.getContext().commands;
+
+    function defaultCompletion(): Arguments | void {
+      const handlers = command.getCommandHandlers();
+      for (let i = 0, ii = args.length; i < ii; ++i) {
+        if (handlers[args[i]] && handlers[args[i]].builder) {
+          const builder = handlers[args[i]].builder;
+          if (isCommandBuilderCallback(builder)) {
+            const y = yargs.reset();
+            builder(y);
+            return y.argv;
+          }
+        }
+      }
+      if (
+        !current.match(/^-/) &&
+        parentCommands[parentCommands.length - 1] !== current
+      ) {
+        usage.getCommands().forEach(usageCommand => {
+          const commandName = parseCommand(usageCommand[0]).cmd;
+          if (args.indexOf(commandName) === -1) {
+            if (!zshShell) {
+              completions.push(commandName);
+            } else {
+              const desc = usageCommand[1] || '';
+              completions.push(commandName.replace(/:/g, '\\:') + ':' + desc);
+            }
+          }
+        });
+      }
+      if (current.match(/^-/) || (current === '' && completions.length === 0)) {
+        const descs = usage.getDescriptions();
+        const options = yargs.getOptions();
+        Object.keys(options.key).forEach(key => {
+          const negable =
+            !!options.configuration['boolean-negation'] &&
+            options.boolean.includes(key);
+          // If the key and its aliases aren't in 'args', add the key to 'completions'
+          let keyAndAliases = [key].concat(aliases[key] || []);
+          if (negable)
```

Comment:
It's this block especially where I think we'd benefit from a bit more decomposition. Could `completeOptionKey` could be pulled out to the top level of the file, along with us breaking apart `defaultCompletion` into a couple more helpers?

I think if we then add comments to each of the helpers, the completion logic might become easier to consume.